### PR TITLE
LibMedia: Create a decoder backed by FFmpeg, various improvements

### PR DESCRIPTION
**Pull request by Zaggy1024**
Using this new FFmpeg decoder, we can now play quite a few more codecs in the Matroska container, including H.264, VP8, VP9, and AV1.

The FFmpeg decoder _should_ be at parity with the Serenity VP9 decoder, so that has been removed in the last commit.

YUV video frames should now consume half the memory for 8-bit videos (which is the vast majority of videos) by storing their data in a byte array and reinterpreting it to u8 or u16 based on the bit depth.

### CODE REVIEW

**Comment by Zaggy1024** [comment]
Marked as draft until I can get this building on the macOS CI. Installing ffmpeg via homebrew doesn't seem to be enough, it probably needs some include directory.

File: Userland/Libraries/LibMedia/Video/VP9/Decoder.cpp [hunk 1]
```diff
@@ -146,36 +156,32 @@ DecoderErrorOr<void> Decoder::create_video_frame(FrameContext const& frame_conte
...
+    auto output_uv_size = subsampling.subsampled_size(output_y_size);
+
+    Array<FixedArray<u8>, 3> output_buffers = {
+        DECODER_TRY_ALLOC(FixedArray<u8>::create(output_y_size.width() * output_y_size.height() * sizeof(T))),
+        DECODER_TRY_ALLOC(FixedArray<u8>::create(output_uv_size.width() * output_uv_size.height() * sizeof(T))),
+        DECODER_TRY_ALLOC(FixedArray<u8>::create(output_uv_size.width() * output_uv_size.height() * sizeof(T))),
     };
     for (u8 plane = 0; plane < 3; plane++) {
-        auto& buffer = output_buffers[plane];
+        auto* buffer = reinterpret_cast<T*>(output_buffers[plane].data());
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/Video/VP9/Decoder.cpp]: any alignment concerns with this data here?

File: Userland/Libraries/LibMedia/VideoFrame.cpp [hunk 2]
```diff
@@ -129,46 +151,60 @@ ALWAYS_INLINE DecoderErrorOr<void> convert_to_bitmap_subsampled(Convert convert,
...
 }
 
-template<u32 subsampling_horizontal, u32 subsampling_vertical>
-static ALWAYS_INLINE DecoderErrorOr<void> convert_to_bitmap_selecting_converter(CodingIndependentCodePoints cicp, u8 bit_depth, u32 const width, u32 const height, FixedArray<u16> const& plane_y, FixedArray<u16> const& plane_u, FixedArray<u16> const& plane_v, Gfx::Bitmap& bitmap)
+template<u32 subsampling_horizontal, u32 subsampling_vertical, typename T>
+static ALWAYS_INLINE DecoderErrorOr<void> convert_to_bitmap_selecting_converter(CodingIndependentCodePoints cicp, u8 bit_depth, u32 const width, u32 const height, ReadonlyBytes plane_y_data, ReadonlyBytes plane_u_data, ReadonlyBytes plane_v_data, Gfx::Bitmap& bitmap)
 {
+    auto const* plane_y = reinterpret_cast<T const*>(plane_y_data.data());
+    auto const* plane_u = reinterpret_cast<T const*>(plane_u_data.data());
+    auto const* plane_v = reinterpret_cast<T const*>(plane_v_data.data());
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/VideoFrame.cpp:161]: same here, re:alignment, any reason not to use AK::ByteReader for this?

File: Userland/Libraries/LibMedia/CMakeLists.txt [hunk 3]
```diff
@@ -15,3 +16,9 @@ set(SOURCES
 
 serenity_lib(LibMedia media)
 target_link_libraries(LibMedia PRIVATE LibCore LibIPC LibGfx LibThreading)
+
+find_package(PkgConfig)
+pkg_check_modules(LIBAV IMPORTED_TARGET libavcodec)
+
+target_include_directories(LibMedia PRIVATE ${LIBAV_INCLUDE_DIRS})
+target_link_libraries(LibMedia PRIVATE ${LIBAV_LIBRARIES})
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/CMakeLists.txt:19]: you should target_link_directories with the library dirs here as well. Assuming that the lib dir is in the default lib path is no bueno.

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegHelpers.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibMedia/CodecID.h>
+
+extern "C" {
+#include <libavcodec/avcodec.h>
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/FFmpeg/FFmpegHelpers.h:12]: ew, does this header really not have extern "C" if __cplusplus in it automagically?

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+#include <LibCore/System.h>
+#include <LibMedia/VideoFrame.h>
+
+#include "FFmpegHelpers.h"
+#include "FFmpegVideoDecoder.h"
+
+namespace Media::FFmpeg {
+
+static AVPixelFormat negotiate_output_format(AVCodecContext*, [[maybe_unused]] AVPixelFormat const* formats)
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:15]: yikes, is this one of those null terminated arrays? does the lib not provide any length info?

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    if (avcodec_open2(codec_context, codec, nullptr) < 0)
+        return DecoderError::format(DecoderErrorCategory::Unknown, "Unknown error occurred when opening FFmpeg codec {}", codec_id);
+
+    auto* packet = av_packet_alloc();
+    if (!packet)
+        return DecoderError::with_description(DecoderErrorCategory::Memory, "Failed to allocate FFmpeg packet"sv);
+
+    auto* frame = av_frame_alloc();
+    if (!frame)
+        return DecoderError::with_description(DecoderErrorCategory::Memory, "Failed to allocate FFmpeg frame"sv);
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:73]: does this error return here leak the context/packet allocated above?

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        return DecoderError::with_description(DecoderErrorCategory::Corrupted, "FFmpeg codec reports that the data is corrupted"sv);
+    }
+}
+
+DecoderErrorOr<NonnullOwnPtr<VideoFrame>> FFmpegVideoDecoder::get_decoded_frame()
+{
+    auto result = avcodec_receive_frame(m_codec_context, m_frame);
+
+    switch (result) {
+    case 0: {
```
Comment (ADKaster) [inline Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:123]: does this have some AV_SUCCESS value we could use instead of 0?

**Review by ADKaster — COMMENTED** [review]
We can add this with the system lib with my one comment about lib dir, but I'd like an issue to remind me to investigate the proper way to tell vpkg how to alternate b/w vcpkg and system libs.

File: Userland/Libraries/LibMedia/Video/VP9/Decoder.cpp (hunk 1 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/Video/VP9/Decoder.cpp]: Hopefully fixed by using aligned_alloc based on the bit depth.

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegHelpers.h (hunk 4 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/FFmpeg/FFmpegHelpers.h:12]: Unfortunately, it appears not. I get undefined references to mangled names without this.

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp (hunk 5 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:15]: They don't mention any alternative in the docs for the function, so I guess this is the best we got.

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp (hunk 7 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:123]: After a cursory check, it appears not. The docs for the function only mention a literal `0`.

File: Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp (hunk 6 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp:73]: Good point, will look into cleaning up the leaks in this later.

Comment (Zaggy1024) [reply Userland/Libraries/LibMedia/FFmpeg/FFmpegVideoDecoder.cpp]: I've added an ArmedScopeGuard to free anything that was successfully allocated.

**Comment by Zaggy1024** [comment]
This should be ready for review, and it would be good if someone could test it on a Mac to see if my CI fiddling actually worked out.

**Comment by maddanio** [comment]
anything special to test? checked out build and ran your branch on my m1 mbp and runs fine I guess...

**Comment by Zaggy1024** [comment]
If video plays back, then it should be good there. Thanks!

**Comment by maddanio** [comment]
Do you have a test site? Youtube doesnt work :)

**Comment by Zaggy1024** [comment]
There are some test files in the Tests/LibMedia directory that should be able to display in the browser. Currently, MSE isn't supported, and the video element can only play file sources.

**Comment by maddanio** [comment]
guess I am late, but the test files work fine

**Comment by maddanio** [comment]
on osx

**Pull request merged**